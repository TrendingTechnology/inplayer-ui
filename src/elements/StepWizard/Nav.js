// @flow
import React from 'react';
import styled, { css } from 'styled-components';
import { ifProp } from 'styled-tools';
import colors from 'config/colors';

const StepNavigation = styled.div`
  margin-bottom: 2rem;
  text-align: center;
`;

const StepAction = styled.span`
  color: black;
  cursor: pointer;
  font-size: 36px;
  line-height: 1;
  margin: 0 15px;
  opacity: 0.4;
  text-shadow: none;
  transition: opacity 1s ease, text-shadow 1s ease;
  will-change: opacity, text-shadow;
  ${ifProp(
    'isActive',
    css`
      color: ${colors.white};
      opacity: 1;
      text-shadow: 0 0px 8px;
    `
  )};
`;

type Props = {
  totalSteps: number,
  currentStep: number,
  goToStep: Function,
};

const Nav = ({ totalSteps, currentStep, goToStep }: Props) => {
  const dots = [];
  for (let i = 1; i <= totalSteps; i += 1) {
    const isActive = currentStep === i;
    dots.push(
      <StepAction isActive={isActive} key={`step-${i}`} onClick={() => goToStep(i)}>
        &bull;
      </StepAction>
    );
  }

  return <StepNavigation>{dots}</StepNavigation>;
};

export default Nav;
