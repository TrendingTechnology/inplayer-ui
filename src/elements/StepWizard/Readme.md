A software wizard or setup assistant is a user interface type that presents a user with a sequence of dialog boxes that lead the user through a series of well-defined steps. Tasks that are complex, infrequently performed, or unfamiliar may be easier to perform using a wizard.

```jsx static
import React from 'react';
import { StepWizard } from '@inplayer-org/inplayer-ui';

function StepOne() {
  return (
    <div>
      <Typography variant="h2">Step One!</Typography>
    </div>
  );
};

function StepTwo() {
  return (
    <div>
      <Typography variant="h2">Step Two!</Typography>
    </div>
  );
};

function StepThree() {
  return (
    <div>
      <Typography variant="h2">Step Three!</Typography>
    </div>
  );
};

function StepFour() {
  return (
    <div>
      <Typography variant="h2">Step Four!</Typography>
    </div>
  );
};


class StepWizardExample extends React.Component {
  constructor() {
    super();

    this.state = {
      activeStep: 0,
    };

    this.onStepChange = this.onStepChange.bind(this);
  }

  onStepChange(activeStep) {
    this.setState({ activeStep });
  }

  render() {
    const { activeStep } = this.state;

    return (
      <StepWizard activeStep={activeStep} onStepChange={onStepChange}>
        <StepOne isDisabled={false} isCompleted={true} />
        <StepTwo isDisabled={false} />
        <StepThree isDisabled={true} />
        <StepFour isDisabled={true} />
      </StepWizard>
    );
  }
}
```

```jsx
function StepOne() {
  return (
    <div>
      <Typography variant="h2">Step One!</Typography>
    </div>
  );
};

function StepTwo() {
  return (
    <div>
      <Typography variant="h2">Step Two!</Typography>
    </div>
  );
};

function StepThree() {
  return (
    <div>
      <Typography variant="h2">Step Three!</Typography>
    </div>
  );
};

function StepFour() {
  return (
    <div>
      <Typography variant="h2">Step Four!</Typography>
    </div>
  );
};


class StepWizardExample extends React.Component {
  constructor() {
    super();

    this.state = {
      activeStep: 0,
    };

    this.onStepChange = this.onStepChange.bind(this);
  }

   onStepChange(activeStep) {
    this.setState({ activeStep });
  }

  render() {
    const { activeStep } = this.state;

    return (
      <React.Fragment>
        <StepWizard activeStep={activeStep} onStepChange={this.onStepChange}>
          <StepOne isDisabled={false} isCompleted={true} />
          <StepTwo isDisabled={false} />
          <StepThree isDisabled={true} />
          <StepFour isDisabled={true} />
        </StepWizard>
      </React.Fragment>
    );
  }
}
<StepWizardExample />;
```
