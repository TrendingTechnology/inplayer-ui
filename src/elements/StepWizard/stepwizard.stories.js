/* eslint-disable no-shadow */
import React, { useState } from 'react';
import styled from 'styled-components';
import { StepWizard, Typography } from 'src/index';

export default {
  component: StepWizard,
  title: 'Step Wizard',
};

const Container = styled.div`
  margin: auto;
  margin-top: 2%;
  width: 90%;
`;

const StepOne = () => (
  <div>
    <Typography variant="h2">Step One</Typography>
  </div>
);
const StepTwo = () => (
  <div>
    <Typography variant="h2">Step Two</Typography>
  </div>
);
const StepThree = () => (
  <div>
    <Typography variant="h2">Step Three</Typography>
  </div>
);
const StepFour = () => (
  <div>
    <Typography variant="h2">Step Four</Typography>
  </div>
);

export const Demo = () => {
  const [activeStep, setActiveStep] = useState(0);

  const onStepChange = (activeStep) => setActiveStep(activeStep);

  const steps = [
    { isDisabled: false, isCompleted: true, component: <StepOne /> },
    { isDisabled: false, isCompleted: false, component: <StepTwo /> },
    { isDisabled: true, isCompleted: false, component: <StepThree /> },
    { isDisabled: true, isCompleted: false, component: <StepFour /> },
  ];

  return (
    <Container>
      <Typography variant="h1" color="black">
        Step Wizard
      </Typography>
      <Typography variant="p" description>
        A software wizard or setup assistant is a user interface type that presents a user with a
        sequence of dialog boxes that lead the user through a series of well-defined steps. Tasks
        that are complex, infrequently performed, or unfamiliar may be easier to perform using a
        wizard.
        <br />
        <br />
        <b>Step three and four are disabled.</b>
      </Typography>

      <StepWizard activeStep={activeStep} onStepChange={onStepChange} steps={steps} />
    </Container>
  );
};
