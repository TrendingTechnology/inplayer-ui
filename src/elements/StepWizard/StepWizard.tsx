// @flow
import React from 'react';
import { CSSProperties } from 'styled-components';
import Nav, { StepItem } from './Nav';
import Step from './Step';

type TransitionVariant = 'fadeInLeft' | 'fadeInRight' | 'fadeOutLeft' | 'fadeOutRight';

type StepWizardProps = {
  /**
   * Array of objects containing steps information and components
   */
  steps: Array<StepItem>;
  /**
   * Active step of step wizard
   */
  activeStep: number;
  /**
   * External CSS class name
   */
  className?: string;
  /**
   * Inline CSS
   */
  style?: CSSProperties;
  /**
   * Function which runs when the step changes
   */
  onStepChange?: (activeStep: number) => void;
  /**
   * CSS transition type
   */
  transition?: TransitionVariant;
};

const StepWizard = ({
  steps,
  activeStep,
  className = '',
  style = {},
  onStepChange = () => {},
  transition = 'fadeInRight',
}: StepWizardProps) => {
  const isInvalidStep = (currentStep: number) => currentStep < 0 || currentStep > steps.length;

  const setActiveStep = (currentStep: number) => {
    if (activeStep === currentStep) return;
    if (isInvalidStep(currentStep)) {
      console.error(`${currentStep} is an invalid step`);
      return;
    }

    // User callback
    onStepChange(currentStep);
  };

  /** Go to step index */
  const goToStep = (step: number) => setActiveStep(step);

  const renderSteps = steps.map((step, i) => {
    const isStepActive = i === activeStep;

    if (isStepActive) {
      return (
        <Step key={`step-${i}`} isActive={isStepActive} transition={transition}>
          {step.component}
        </Step>
      );
    }

    return null;
  });

  return (
    <div className={className} style={style}>
      <Nav steps={steps} activeStep={activeStep} goToStep={goToStep} />
      <div>{renderSteps}</div>
    </div>
  );
};

export default StepWizard;
