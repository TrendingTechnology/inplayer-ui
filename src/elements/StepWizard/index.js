// @flow
import React from 'react';
import styled from 'styled-components';
import Nav from './Nav';
import Step from './Step';

type TransitionVariant = 'fadeInLeft' | 'fadeInRight' | 'fadeOutLeft' | 'fadeOutRight';

type Props = {
  children: React.ReactNode,
  activeStep: number,
  className?: string,
  style?: Object,
  onStepChange?: (activeStep: number) => void,
  transition?: TransitionVariant,
};

const WizardContainer = styled.div`
  position: relative;
`;

const StepWizard = ({
  children,
  className,
  style,
  activeStep,
  onStepChange,
  transition,
}: Props) => {
  const isInvalidStep = currentStep => currentStep < 0 || currentStep > children.length;

  const setActiveStep = currentStep => {
    if (activeStep === currentStep) return;
    if (isInvalidStep(currentStep)) {
      console.error(`${currentStep} is an invalid step`);
      return;
    }

    // User callback
    onStepChange(currentStep);
  };

  /** Go to step index */
  const goToStep = step => setActiveStep(step);

  const childrenWithProps = React.Children.map(children, (child, i) => {
    // child has 0 starting index
    const isStepActive = i + 1 === activeStep;

    if (isStepActive) {
      return (
        <Step isActive={isStepActive} transition={transition}>
          {child}
        </Step>
      );
    }

    return null;
  });

  return (
    <div className={className} style={style}>
      <Nav activeStep={activeStep} totalSteps={children.length} goToStep={goToStep} />
      <WizardContainer>{childrenWithProps}</WizardContainer>
    </div>
  );
};

StepWizard.defaultProps = {
  className: null,
  style: {},
  transition: 'fadeInRight',
  onStepChange: () => {},
};

export default StepWizard;
