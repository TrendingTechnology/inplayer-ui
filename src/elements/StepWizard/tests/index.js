import React from 'react';
import { generateSnapshots } from 'helpers';

import StepWizard from '../index';

const StepOne = () => (
  <div>
    <h1 variant="h2">Step One!</h1>
  </div>
);

const StepTwo = () => (
  <div>
    <h1 variant="h2">Step Two!</h1>
  </div>
);

const StepThree = () => (
  <div>
    <h1 variant="h2">Step Three!</h1>
  </div>
);

const StepFour = () => (
  <div>
    <h1 variant="h2">Step Four!</h1>
  </div>
);

const StepWizardComponent = () => (
  <StepWizard>
    <StepOne />
    <StepTwo />
    <StepThree />
    <StepFour />
  </StepWizard>
);

generateSnapshots(StepWizardComponent);
