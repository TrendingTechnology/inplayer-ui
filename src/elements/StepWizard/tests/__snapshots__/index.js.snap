// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepWizardComponent renders correctly 1`] = `
.c0 {
  margin-bottom: 2rem;
  text-align: center;
}

.c1 {
  color: black;
  cursor: pointer;
  font-size: 36px;
  line-height: 1;
  margin: 0 15px;
  opacity: 0.4;
  text-shadow: none;
  -webkit-transition: opacity 1s ease,text-shadow 1s ease;
  transition: opacity 1s ease,text-shadow 1s ease;
  will-change: opacity,text-shadow;
}

.c2 {
  position: relative;
}

<div
  className={null}
  style={Object {}}
>
  <div
    className="c0"
  >
    <span
      className="c1"
      onClick={[Function]}
    >
      •
    </span>
    <span
      className="c1"
      onClick={[Function]}
    >
      •
    </span>
    <span
      className="c1"
      onClick={[Function]}
    >
      •
    </span>
    <span
      className="c1"
      onClick={[Function]}
    >
      •
    </span>
  </div>
  <div
    className="c2"
  />
</div>
`;
